File: src/07.greedy/02.subArrayGreedy/16.85.ts
export function maximalRectangle(matrix: string[][]): number {
    /**
     // 暴力解法
    // 从左往右, 记录每个点连续的宽度
    // 从下往上, 记录每个点连续的高度

    const size: number[][] = Array.from(matrix).map(() =>
        new Array(matrix[0].length).fill(0)
    );

    let maxArea = 0;

    for (let i = 0; i < matrix.length; i++) {
        for (let j = 0; j < matrix[0].length; j++) {
            // 更新 width
            if (matrix[i][j] === "1") {
                if (j === 0) {
                    size[i][j] = 1;
                } else {
                    size[i][j] = size[i][j - 1] + 1;
                }
            } else {
                size[i][j] = 0;
            }

            // 当前列中, 所有行的最小宽度
            let minWidth = size[i][j];

            // 向上扩展
            for (let k = i; k >= 0; k--) {
                const height = i - k + 1;
                minWidth = Math.min(minWidth, size[k][j]);
                maxArea = Math.max(maxArea, height * minWidth);
            }
        }
    }

    return maxArea;
     */

    // 利用 84 题, 求最高矩形的面积计算方法, 单调栈
    const heights = new Array(matrix[0].length + 2).fill(0);
    const stack = [];
    let maxArea = 0;

    for (let i = 0; i < matrix.length; i++) {
        for (let j = 0; j < heights.length; j++) {
            if (
                j === 0 ||
                j === heights.length - 1 ||
                matrix[i][j - 1] === "0"
            ) {
                heights[j] = 0;
            } else {
                heights[j] += 1;
            }

            while (
                stack.length &&
                heights[j] < heights[stack[stack.length - 1]]
            ) {
                const cur = stack.pop();
                const leftMin = stack[stack.length - 1];

                // 右侧边界, 不包含
                const rightMin = j - 1;
                const area = (rightMin - leftMin) * heights[cur];
                maxArea = Math.max(maxArea, area);
            }
            stack.push(j);
        }
    }

    return maxArea;
}

File: src/07.greedy/02.subArrayGreedy/16.85.go
package sub_array_greedy

func MaximalRectangle(matrix [][]byte) int {
	hLen := len(matrix[0]) + 2
	heights := make([]int, hLen)
	stack := []int{}
	maxArea := 0

	for _, row := range matrix {
		for j := 0; j < hLen; j++ {

			// 动态计算 矩形高度
			if j == 0 || j == hLen-1 || row[j-1] == '0' {
				heights[j] = 0
			} else {
				heights[j] += 1
			}

			// 判断是否找到最高柱子的索引
			for len(stack) > 0 && heights[stack[len(stack)-1]] > heights[j] {
				cur := stack[len(stack)-1]

				// 此时, j 为当前最高柱子的右边界, 不包含
				right := j - 1

				// 左边界为: 栈顶元素的下一个元素
				stack = stack[:len(stack)-1]
				left := stack[len(stack)-1]

				// 确定面积
				area := (right - left) * heights[cur]

				// 计算最大面积

				if maxArea < area {
					maxArea = area
				}
			}

			stack = append(stack, j)
		}
	}

	return maxArea
}
